tests/e2e/json_utils.go

```diff
@@ -0,0 +1,448 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+)
+
+// MarshalJSON marshals a step into JSON while including the type of the action.
+func (step Step) MarshalJSON() ([]byte, error) {
+	actionType := reflect.TypeOf(step.Action)
+
+	result := struct {
+		ActionType string
+		Action     interface{}
+		State      State
+	}{
+		ActionType: actionType.String(),
+		Action:     step.Action,
+		State:      step.State,
+	}
+
+	return json.Marshal(result)
+}
+
+// UnmarshalJSON unmarshals a step from JSON while including the type of the action.
+func (step *Step) UnmarshalJSON(data []byte) error {
+	var tmp struct {
+		ActionType string
+		Action     json.RawMessage
+		State      State
+	}
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+
+	action, err := UnmarshalMapToActionType(tmp.Action, tmp.ActionType)
+	if err != nil {
+		return err
+	}
+
+	step.Action = action
+	step.State = tmp.State
+	return nil
+}
+
+// UnmarshalMapToActionType takes a JSON object and an action type and marshals into an object of the corresponding action.
+func UnmarshalMapToActionType(rawAction json.RawMessage, actionTypeString string) (interface{}, error) {
+	switch actionTypeString {
+	case "main.submitConsumerAdditionProposalAction":
```

Comment:
Is there a reason some of these actions are exported/capitalized and some are not?

Reply:
These need to be the way the types are named in the go code. This is just continuing the existing inconsistency.